src/graphnet/models/gnn/icemix.py

```diff
@@ -0,0 +1,240 @@
+"""Implementation of IceMix architecture used in.
+
+                    IceCube - Neutrinos in Deep Ice
+Reconstruct the direction of neutrinos from the Universe to the South Pole
+
+Kaggle competition.
+
+Solution by DrHB: https://github.com/DrHB/icecube-2nd-place
+"""
+import torch
+import torch.nn as nn
+from typing import List, Set
+
+from graphnet.models.components.layers import (
+    FourierEncoder,
+    SpacetimeEncoder,
+    Block_rel,
+    Block,
+)
+from graphnet.models.gnn.dynedge import DynEdge
+from graphnet.models.gnn.gnn import GNN
+
+from torch_geometric.utils import to_dense_batch
+from torch_geometric.data import Data
+from torch import Tensor
+
+
```

Comment:
I think this is a really cool function and that it could be a valuable utility function outside the context of this new model. 

I've refactored the function slightly and added additional doc strings to make it clearer what it actually does. We could consider moving this to `graphnet.models.utils.py` instead of having it in this file.

 Please see below a suggestion on how this function could look like. 

```python
def array_to_sequence(x: Tensor, 
                      batch_idx: LongTensor, 
                      padding_value: Tensor = torch.inf) -> Tuple[Tensor, 
                                                                  Tensor, 
                                                                  Tensor]:
    """Convert `x` of shape [n,d] into a padded sequence of shape [B, L, D].

       Where B is the batch size, L is the sequence length and D is the 
       features for each time step.

    Args:
        x: array-like tensor with shape `[n,d]` where `n` is the total number 
        of pulses in the batch and `d` is the number of  node features.
        batch_idx: a LongTensor identifying which row in `x` belongs to
                 which training example. 
                 E.g. `torch_geometric.data.Batch.batch`.
    Returns:
        x: Padded sequence with dimensions  [B, L, D].
        mask: A tensor that identifies masked entries in `x`.
               E.g. : `masked_entries = x[mask]`
        seq_length: A tensor containing the number of pulses in each event.
    """
 
    _, seq_length = torch.unique(batch_idx, return_counts=True)
    x_list = torch.split(x, seq_length.tolist())

    x = torch.nn.utils.rnn.pad_sequence(
        x_list, batch_first=True, padding_value=padding_value
    )
    mask = torch.ne(x[:, :, 1], torch.inf)
    x[~mask] = 0
    return x, mask, seq_length
```

Reply:
Nice comment, I think this is a better way to structure the function. However, I am a but concern about the padding value.
However, if the user chooses to specify a padding value, it should look like `x[~mask] = padding_value`, since this padding_value is the value that will be used to pad the output Tensor x.
For the intermediate step I chosed `torch.inf` to avoid false possitives when using `torch.ne`.
Some pseudo-code would be like this, let me know your comments:

```python
def array_to_sequence(x: Tensor, 
                      batch_idx: LongTensor, 
                      padding_value: Any = 0,
                      excluding_value: Any = torch.inf) -> Tuple[Tensor, 
                                                                  Tensor, 
                                                                  Tensor]:
    """Convert `x` of shape [n,d] into a padded sequence of shape [B, L, D].

       Where B is the batch size, L is the sequence length and D is the 
       features for each time step.

    Args:
        x: array-like tensor with shape `[n,d]` where `n` is the total number 
        of pulses in the batch and `d` is the number of  node features.
        batch_idx: a LongTensor identifying which row in `x` belongs to
                which training example. 
                E.g. `torch_geometric.data.Batch.batch`.
        padding_value: The value to use for padding.
        excluding_value: This parameter represents a unique value that should
                not be present in the input tensor 'x'
    Returns:
        x: Padded sequence with dimensions  [B, L, D].
        mask: A tensor that identifies masked entries in `x`.
               E.g. : `masked_entries = x[mask]`
        seq_length: A tensor containing the number of pulses in each event.
    """
    
    if torch.any(torch.eq(x, excluding_value)):
        raise ValueError(f'Transformation cannot be made because input tensor '
                         f'`x` contains at least one element equal to '
                         f'excluding value {excluding_value}.')
    
    _, seq_length = torch.unique(batch_idx, return_counts=True)
    x_list = torch.split(x, seq_length.tolist())
    
    x = torch.nn.utils.rnn.pad_sequence(
        x_list, batch_first=True, padding_value=excluding_value
    )
    mask = torch.ne(x[:, :, 1], excluding_value)
    x[~mask] = padding_value
    return x, mask, seq_length
```